packages/fxa-settings/src/pages/Signin/SigninTotpCode/container.tsx

```diff
@@ -0,0 +1,111 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import { RouteComponentProps, useNavigate, useLocation } from '@reach/router';
+import { useValidatedQueryParams } from '../../../lib/hooks/useValidate';
+import { SigninQueryParams } from '../../../models/pages/signin';
+import { SigninTotpCode } from './index';
+import { gql, useMutation } from '@apollo/client';
+import GleanMetrics from '../../../lib/glean';
+import { logViewEvent } from '../../../lib/metrics';
+import { hardNavigate } from 'fxa-react/lib/utils';
+import { AuthUiErrors } from '../../../lib/auth-errors/auth-errors';
+import { currentAccount, sessionToken } from '../../../lib/cache';
+import { MozServices } from '../../../lib/types';
+import VerificationMethods from '../../../constants/verification-methods';
+
+export const viewName = 'signin-totp-code';
+
+export type SigninTotpCodeContainerProps = {
+  serviceName: MozServices;
+};
+
+export const SignInTotpCodeContainer = ({
+  serviceName,
+}: SigninTotpCodeContainerProps & RouteComponentProps) => {
+  const navigate = useNavigate();
+  const location = useLocation();
+  const storedLocalAccount = currentAccount();
+
+  // TBD: A bit of a work around. In the previous signup step
+  // a new session token should have been generated. Setting
+  // it here to ensure that the authorization header is set
+  // when it comes time to verify the totp code.
+  if (storedLocalAccount?.sessionToken) {
+    sessionToken(storedLocalAccount?.sessionToken);
+  }
+
+  const { queryParamModel } = useValidatedQueryParams(SigninQueryParams);
+  const [verifyTotpCode] = useMutation(gql`
+    mutation VerifyTotp($input: VerifyTotpInput!) {
+      verifyTotp(input: $input) {
+        success
+      }
+    }
+  `);
+
+  const handleNavigation = () => {
+    console.log('!!! handleNavigation', queryParamModel);
+
+    // Note: `redirectTo` looks similar to `redirectPathname`, but they
+    // work slightly differently: `redirectTo` redirects automatically,
+    // while `redirectPathname` redirects after a form submission.
+    if (queryParamModel.redirectToUrl) {
+      hardNavigate(queryParamModel.redirectToUrl);
+    } else if (queryParamModel.redirectPathname) {
+      navigate(queryParamModel.redirectPathname + location.search);
+    } else if (storedLocalAccount?.verificationReason === 'change_password') {
```

Comment:
Should this check supersede the other two? (if the verification reason is "change_password", navigate to force_password_change even if there is a redirectToUrl or redirectPathname)?

Reply:
I have a related comment below I can't link to since I haven't posted it yet.

Also, this can reference the `verificationReasons` enum.

Reply:
I didn't actually scrutinize this. I copied the logic (here)[https://github.com/mozilla/fxa/blob/FXA-6491/packages/fxa-content-server/app/scripts/views/sign_in_token_code.js#L104]. But what you are saying seems to make sense to me... @LZoog, what do think about this...

Reply:
Is the `redirectPathname` related to OAuth at all? I'll dig in there a little bit to see what this is doing...

Reply:
Okay, I just looked into it. It seems like `redirectPathname` is used in the backbone version to get back to whatever step you were on prior to redirecting to totp or sign code verification. It appears to be used by the upgrade-session-redirect mixin, which is automatically hooks into form  submission and handles UNVERIFIED_SESSION errors. So in a nut shell it's a redirect back into the auth flow.

Reply:
@dschom I see `redirectPathname` was introduced here: https://github.com/mozilla/fxa/pull/6295/files

I don't think this is a param that RPs use, it looks to me like we had a case in "old settings" where after a user went through the pairing flow, they could not delete their account because the new session wasn't verified. In the `upgrade-sessino-redirect-mixin`, it looks like we [pass this value to the page](https://github.com/mozilla/fxa/blob/main/packages/fxa-content-server/app/scripts/views/mixins/upgrade-session-redirect-mixin.js#L35-L36) (unsure how that sets it on the model without digging). I'm pretty sure this is addressed in new Settings via our `VerifiedSessionGuard` around the action - I don't think we need to account for this. What do you think?

Reply:
I was also thinking we probably don't need it, but... I am not sure I'd remove it. There's enough stuff where backbone and react might be mixed together that I'd be hesitant to take it out. Maybe we should query the server logs and validate we don't use it before removing it?